Detectors/TRD/simulation/src/Detector.cxx

```diff
@@ -121,48 +134,113 @@ bool Detector::ProcessHits(FairVolume* v)
     fMC->TrackPosition(x, y, z);
-    double time = fMC->TrackTime() * 1e6;
+    double tof = fMC->TrackTime() * 1e6; // The time of flight in micro-seconds
     o2::Data::Stack* stack = (o2::Data::Stack*)fMC->GetStack();
     const int trackID = stack->GetCurrentTrackNumber();
-    addHit(x, y, z, time, enDep, trackID, det);
+    addHit(x, y, z, tof, enDep, trackID, det);
     stack->addHit(GetDetId());
     return true;
   }
   return false;
 }
 
+void Detector::CreateTRhit(int det)
+{
+  //
+  // Creates an electron cluster from a TR photon.
+  // The photon is assumed to be created a the end of the radiator. The
+  // distance after which it deposits its energy takes into account the
+  // absorbtion of the entrance window and of the gas mixture in drift
+  // volume.
+  //
+
+  // Maximum number of TR photons per track
+  constexpr int mMaxNumberOfTRPhotons = 50; // Make this a class member?
+
+  TLorentzVector mom;
+  fMC->TrackMomentum(mom);
+  float pTot = mom.Rho();
+  std::vector<float> photonEnergyContainer;            // energy in keV
+  mTR->createPhotons(11, pTot, photonEnergyContainer); // Create TR photons
+  if (photonEnergyContainer.size() > mMaxNumberOfTRPhotons) {
+    LOG(FATAL) << "Boundary error: nTR = " << photonEnergyContainer.size() << ", mMaxNumberOfTRPhotons = " << mMaxNumberOfTRPhotons;
+  }
+
+  // Loop through the TR photons
+  for (const float& photonEnergy : photonEnergyContainer) {
+    const double energyMeV = photonEnergy * 1e-3;
+    const double energyeV = photonEnergy * 1e3;
+    double absLength = 0.0;
+    double sigma = 0.0;
+    // Take the absorbtion in the entrance window into account
+    double muMy = mTR->getMuMy(energyMeV);
+    sigma = muMy * mFoilDensity;
+    if (sigma > 0.0) {
+      absLength = gRandom->Exp(1.0 / sigma);
+      if (absLength < TRDGeometry::myThick()) {
+        continue;
+      }
+    } else {
+      continue;
+    }
+    // The absorbtion cross sections in the drift gas
+    // Gas-mixture (Xe/CO2)
+    double muNo = 0.0;
+    if (TRDCommonParam::Instance()->IsXenon()) {
+      muNo = mTR->getMuXe(energyMeV);
+    } else if (TRDCommonParam::Instance()->IsArgon()) {
+      muNo = mTR->getMuAr(energyMeV);
+    }
+    double muCO = mTR->getMuCO(energyMeV);
+    double fGasNobleFraction = 1;
+    double fGasDensity = 1;
+    sigma = (fGasNobleFraction * muNo + (1.0 - fGasNobleFraction) * muCO) * fGasDensity * mTR->getTemp();
+
+    // The distance after which the energy of the TR photon
+    // is deposited.
+    if (sigma > 0.0) {
+      absLength = gRandom->Exp(1.0 / sigma);
+      if (absLength > (TRDGeometry::drThick() + TRDGeometry::amThick())) {
+        continue;
+      }
+    } else {
+      continue;
+    }
+
+    // The position of the absorbtion
+    TLorentzVector pos;
+    fMC->TrackPosition(pos);
+    double x = pos[0] + mom[0] / pTot * absLength;
+    double y = pos[1] + mom[1] / pTot * absLength;
+    double z = pos[2] + mom[2] / pTot * absLength;
```

Comment:
we are not using "fast-math" so these divisions will not be optimized away. I suggest to do
`auto invpTot = 1./pTot`; and then `mom[2]*invpTot`;

Moreover, do you need doubles or would float be ok?

Reply:
I see the point. I am changing the arithmetics.

For the type of the coordinates I will create another branch where I will fix this in other places too.

Reply:
@sawenzel, I confirm that there are a number of places where the use of double instead of float in unnecessary. As I proposed I will change this in a new branch to avoid problems with #1649.